agent/dns.go

```diff
@@ -1483,10 +1369,284 @@ func findWeight(node structs.CheckServiceNode) int {
 	}
 }
 
+// serviceNodeAddr is used to identify target service address
+func (d *DNSServer) serviceNodeAddr(serviceNode structs.CheckServiceNode, dc string, dnsQuery string) string {
+	addr := d.agent.TranslateAddress(dc, serviceNode.Node.Address, serviceNode.Node.TaggedAddresses)
+
+	if serviceNode.Service.Address != "" {
+		addr = serviceNode.Service.Address
+	}
+
+	// If the service address is a CNAME for the service we are looking
+	// for then use the node address.
+	if dnsQuery == strings.TrimSuffix(addr, ".")+"." {
+		addr = serviceNode.Node.Address
+	}
+
+	return addr
+}
+
+func (d *DNSServer) encodeIPAsFqdn(dc, ip string) string {
+	addr := net.ParseIP(ip)
+	ipv4 := addr.To4()
+	if ipv4 != nil {
+		ipStr := hex.EncodeToString(addr)
+		return fmt.Sprintf("%s.addr.%s.%s", ipStr[len(ipStr)-(net.IPv4len*2):], dc, d.domain)
+	} else {
+		return fmt.Sprintf("%s.addr.%s.%s", hex.EncodeToString(addr), dc, d.domain)
+	}
+}
+
+func makeARecord(qType uint16, addr string, ttl time.Duration) dns.RR {
+	ip := net.ParseIP(addr)
+
+	var ipRecord dns.RR
+	ipv4 := ip.To4()
+	if ipv4 != nil {
+		if qType == dns.TypeSRV || qType == dns.TypeA || qType == dns.TypeANY || qType == dns.TypeNS || qType == dns.TypeTXT {
+			ipRecord = &dns.A{
+				Hdr: dns.RR_Header{
+					Rrtype: dns.TypeA,
+					Class:  dns.ClassINET,
+					Ttl:    uint32(ttl / time.Second),
+				},
+				A: ipv4,
+			}
+		}
+	} else if qType == dns.TypeSRV || qType == dns.TypeAAAA || qType == dns.TypeANY || qType == dns.TypeNS || qType == dns.TypeTXT {
+		ipRecord = &dns.AAAA{
+			Hdr: dns.RR_Header{
+				Rrtype: dns.TypeAAAA,
+				Class:  dns.ClassINET,
+				Ttl:    uint32(ttl / time.Second),
+			},
+			AAAA: ip,
+		}
+	}
+	return ipRecord
+}
+
+// Craft dns records for a node
+// In case of an SRV query the answer will be a IN SRV and additional data will store an IN A to the node IP
+// Otherwise it will return a IN A record
+func (d *DNSServer) makeRecordFromNode(dc string, node *structs.Node, qType uint16, qName string, ttl time.Duration, maxRecursionLevel int) []dns.RR {
+	addr := d.agent.TranslateAddress(dc, node.Address, node.TaggedAddresses)
+
+	var res []dns.RR
+
+	if net.ParseIP(addr) == nil {
```

Comment:
We are going to parse the IP here and potentially again in the call to `makeARecord` below. If we moved the `ParseIP` call out of the `if` statement then we could just pass a `net.IP` to `makeARecord` instead of redoing the parsing within that function.